flax/cursor.py

```diff
@@ -49,97 +51,138 @@ def is_named_tuple(obj):
+    tuple key path will always be non-empty as well.
+  - If ``cond_fn`` is not None, then ``Cursor.find_all`` is calling this
+    function and ``_traverse_tree`` will return a generator where each
+    generated element is of type Tuple[Union[str, int], AccessType]. The
+    tuple contains the key path and access type where the object was found
+    that fulfilled the conditions of the ``cond_fn``.
+  """
+  if (update_fn and cond_fn) or (update_fn is None and cond_fn is None):
+    raise TraverseTreeError(update_fn, cond_fn)
+
   if path:
     str_path = '/'.join(str(key) for key, _ in path)
-    new_obj = update_fn(str_path, obj)
-    if new_obj is not obj:
-      yield path, new_obj
+    if update_fn:
+      new_obj = update_fn(str_path, obj)
+      if new_obj is not obj:
+        yield path, new_obj
+        return
+    elif cond_fn(str_path, obj):  # type: ignore
+      yield path
       return
 
   if isinstance(obj, (FrozenDict, dict)):
     items = obj.items()
-    access_type = AccessType.GETITEM
+    access_type = AccessType.ITEM
   elif is_named_tuple(obj):
     items = ((name, getattr(obj, name)) for name in obj._fields)  # type: ignore
-    access_type = AccessType.GETATTR
+    access_type = AccessType.ATTR
   elif isinstance(obj, (list, tuple)):
     items = enumerate(obj)
-    access_type = AccessType.GETITEM
+    access_type = AccessType.ITEM
   elif dataclasses.is_dataclass(obj):
     items = (
         (f.name, getattr(obj, f.name))
         for f in dataclasses.fields(obj)
         if f.init
     )
-    access_type = AccessType.GETATTR
+    access_type = AccessType.ATTR
   else:
     yield from ()  # empty generator
     return
 
-  for key, value in items:
-    yield from _get_changes(path + ((key, access_type),), value, update_fn)
+  if update_fn:
+    for key, value in items:
+      yield from _traverse_tree(
+          path + ((key, access_type),), value, update_fn=update_fn
+      )
+  else:
+    for key, value in items:
+      yield from _traverse_tree(
+          path + ((key, access_type),), value, cond_fn=cond_fn
+      )
 
 
 class Cursor(Generic[A]):
-  obj: A
-  parent_key: Optional[ParentKey[A]]
-  changes: Dict[Any, Union[Any, 'Cursor[A]']]
+  _obj: A
+  _parent_key: Optional[ParentKey[A]]
+  _changes: Dict[Any, 'Cursor[A]']
 
   def __init__(self, obj: A, parent_key: Optional[ParentKey[A]]):
     # NOTE: we use `vars` here to avoid calling `__setattr__`
     # vars(self) = self.__dict__
-    vars(self)['obj'] = obj
-    vars(self)['parent_key'] = parent_key
-    vars(self)['changes'] = {}
+    vars(self)['_obj'] = obj
+    vars(self)['_parent_key'] = parent_key
+    vars(self)['_changes'] = {}
 
   @property
   def root(self) -> 'Cursor[A]':
-    if self.parent_key is None:
+    if self._parent_key is None:
       return self
     else:
-      return self.parent_key.parent.root  # type: ignore
+      return self._parent_key.parent.root  # type: ignore
+
+  @property
+  def path(self) -> str:
```

Comment:
I think we should expose the minimum amount of attributes because they act as reserved names the user cannot use on their objects. We should make `root` and `path` private (`_root`, `_path`).